File: 프로그래머스/lv1/118666. 성격 유형 검사하기/성격 유형 검사하기.cpp
#include <string>
#include <vector>

using namespace std;

pair<int, int> RT;
pair<int, int> CF;
pair<int, int> JM;
pair<int, int> AN;

int check(int choices)
{
    int a = 0;
    if (choices == 1) a = 3;
    else if (choices == 2) a = 2;
    else if (choices == 3) a = 1;
    return a;
}

string solution(vector<string> survey, vector<int> choices)
{
    string answer = "";

    for (int i = 0; i < survey.size(); i++)
    {
        if (survey[i] == "RT" || survey[i] == "TR")
        {
            if (survey[i] == "RT")
                choices[i] > 4 ? RT.second += choices[i] % 4 : RT.first += check(choices[i]);
            else
                choices[i] > 4 ? RT.first += choices[i] % 4 : RT.second += check(choices[i]);
        }
        else if (survey[i] == "CF" || survey[i] == "FC")
        {
            if (survey[i] == "CF")
                choices[i] > 4 ? CF.second += choices[i] % 4 : CF.first += check(choices[i]);
            else
                choices[i] > 4 ? CF.first += choices[i] % 4 : CF.second += check(choices[i]);
        }
        else if (survey[i] == "JM" || survey[i] == "MJ")
        {
            if (survey[i] == "JM")
                choices[i] > 4 ? JM.second += choices[i] % 4 : JM.first += check(choices[i]);
            else
                choices[i] > 4 ? JM.first += choices[i] % 4 : JM.second += check(choices[i]);
        }
        else if (survey[i] == "AN" || survey[i] == "NA")
        {
            if (survey[i] == "AN")
                choices[i] > 4 ? AN.second += choices[i] % 4 : AN.first += check(choices[i]);
            else
                choices[i] > 4 ? AN.first += choices[i] % 4 : AN.second += check(choices[i]);
        }
    }
    if (RT.first < RT.second)
        answer += "T";
    else
        answer += "R";
    if (CF.first < CF.second)
        answer += "F";
    else
        answer += "C";
    if (JM.first < JM.second)    
        answer += "M";
    else
        answer += "J";
    if (AN.first < AN.second)
        answer += "N";
    else
        answer += "A";
    
    return answer;
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.*;

class Solution {
    
    int[] points = {0, 3, 2, 1, 0, 1, 2, 3};
    String[] types = {"", "RT", "CF", "JM", "AN"};
    
    public String solution(String[] survey, int[] choices) {
        Map<Integer, int[]> map = new TreeMap<>();
        
        for (int i = 0; i < survey.length; i++) {
            char[] type = survey[i].toCharArray();
            int key = type[0] + type[1];
            int[] scores = map.getOrDefault(key, new int[91]);
            
            this.check(type, choices[i], scores);
            map.put(type[0] + type[1], scores);
        }
        
        StringBuilder sb = new StringBuilder();
        for (int i = 1; i < 5; i++) {
            String type = types[i];
            int key = type.charAt(0) + type.charAt(1);
            
            if (map.containsKey(key)) {
                int[] scores = map.get(key);
                int first = scores[type.charAt(0)];
                int second = scores[type.charAt(1)];
                if (first > second) {
                    sb.append(type.charAt(0));
                } else if (first < second) {
                    sb.append(type.charAt(1));
                } else {
                    char max = (char)Math.min(type.charAt(0), type.charAt(1));
                    sb.append(max);
                }
            } else {
                char max = (char)Math.min(type.charAt(0), type.charAt(1));
                sb.append(max);
            }
        }
        
        return sb.toString();
    }
    
    private void check(char[] type, int choice, int[] scores) {
        int point = points[choice];
        if (choice < 4) {
            scores[type[0]] += point;
        } else if (choice > 4) {
            scores[type[1]] += point;
        }
    }
}